# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [22]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [23]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [25]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [26]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [27]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [28]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [29]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [30]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [31]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [32]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [33]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [34]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [35]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [36]:
data.duplicated().sum()

71

In [37]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [38]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [39]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  
0            253875                 покупка жилья                     B  
1            112080       приобретение автомобиля                     C  
2            145885                 покупка жилья                     C  
3            267628    дополнительное образование                     B  
4            158616               сыграть свадьбу                     C  
...             ...                           ...                   ...  
21520        224791             операции с жильем                     B  
21521        155999          сделка с автомобилем                     C  
21522         89672                  недвижимость                     C  
21523        244093  на покупку своего автомобиля                     B  
21524         82047         на покупку автомобиля                     C  

[21331 rows x 13 columns]

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [40]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [41]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [42]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data


children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

<div class='alert alert-info'> Для выявления зависимости между количеством детей и возвратом кредита в срок создаем функцию, которая возвращает категорию по наличию и количесвту детей. Затем можно сгруппировать категории заемщиков с долгом, получить с помощью агрегационных функций общее количество заемщиков, среднее значение по долгу и общее количество просроченных платежей.</div>

In [43]:
#создаем функцию, которая категоризирует данные по количеству детей
def children_category (row):
    children = row['children'] 
    if children == 0:
        return "нет детей"
    elif children <= 2:
        return "малодетные"
    elif children >= 3:
        return "многодетные"

#формируем новый столбец со значением-резульатом работы функции children_category
data['children_category'] = data.apply(children_category, axis=1)   
    
#группируем столбец 'children_category' по столбцу 'debt' и применяем агрегирующие функции к нему (общее количество детей,
#среднее значение просроченных платежей и общее количество просроченных платежей
debtors_with_children = data.groupby('children_category')['debt'].agg(['count', 'mean', 'sum'])

#сортируем по общему количеству детей по убыванию
debtors_with_children = debtors_with_children.sort_values(['count'], ascending=False)

#преобразовываем среднее значение в проценты и округляем до двух знаков после точки
debtors_with_children['mean'] = (debtors_with_children['mean'] * 100).round(2)

debtors_with_children

                                                            

count  mean   sum
children_category                   
нет детей          14091  7.54  1063
малодетные          6860  9.30   638
многодетные          380  8.16    31

<div style='border:solid green 2px; pdding:20px'>
    
    
**Вывод:** 
- Заемщиков без детей больше всего (14091), также у них наименьшая средняя сумма долга от общей суммы (7.54%). Это может быть связано тем, что у бездетных людей более стабильное финансовое положение.
- Общая сумма у заемщиков без детей значительно выше тех, у кого есть дети (1063).
- Среди заемщиков с детьми, малодетные семьи имеют более высокий уровень задолженности, чем многодетные семьи. Это может быть связано с тем, что малодетные семьи могут иметь более высокий уровень доходов и могут брать кредиты на более крупные суммы, чем многодетные семьи.


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

<div class='alert alert-info'>Для выявления зависимости между семейным положением и возвратом кредита в срок группируем 'family_status' с 'debt', далее применяем агрегирующие функции к нему (общее количество заемщиков, среднее значение просроченных платежей и общее количество просроченных платежей)</div>

In [44]:
#группируем столбец 'family_status' по столбцу 'debt' и применяем агрегирующие функции к нему 
debtors_family_status = data.groupby('family_status')['debt'].agg(['count', 'mean', 'sum'])

#преобразовываем среднее значение в проценты и округляем до двух знаков после точки
debtors_family_status['mean'] = (debtors_family_status['mean'] * 100).round(2)

#сортируем по общему количесвту должников по убыванию
debtors_family_status = debtors_family_status.sort_values(['count'], ascending=False)


debtors_family_status



count  mean  sum
family_status                          
женат / замужем        12261  7.56  927
гражданский брак        4134  9.31  385
Не женат / не замужем   2796  9.76  273
в разводе               1189  7.06   84
вдовец / вдова           951  6.62   63

<div style='border:solid green 2px; pdding:20px'>

**Вывод:** 
- Количество женатых заемщиков значительно больше, чем других категорий (12261). Это может быть объяснено тем, что женатые люди имеют стабильное финансовое положение. 
- Процент просрочки кредита больше у неженатых и незамужних заемщиков (9.76%), а у женатых 7.56%.  Может иметь значение то, что женатые люди имеют больше обязательств.
- Процент и количество просроченных кредитов меньше всего у заемщиков из категории 'вдовец/вдова' (6.62% и 63).
- Влияние на оплату кредита в срок также влияет то, что состоящие или состоявшие в браке 
более финансово дисциплинированы


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

<div class='alert alert-info'>Для выявления зависимости между уровнем дохода и возвратом кредита в срок создаем функции, которая возвращает категории по размеру заработной платы. Применяем агрегирующие функции к нему (общее количество должников, среднее значение просроченных платежей и общее количество просроченных платежей)</div>

In [26]:
#создаем функцию, которая категоризирует данные по размеру заработной платы
def salary_category(row):
    total_income = row['total_income'] 
    if total_income < 80000:
        return "средний"
    elif total_income < 150000:
        return "высокий"
    else:
        return "очень высокий"
    
#формируем новый столбец со значением-резульатом работы функции salary_category
data['total_income_category'] = data.apply(salary_category, axis=1)   

#группируем столбец 'total_income_category' по столбцу 'debt' и применяем агрегирующие функции к нему (общее количество должников,
#среднее значение просроченных платежей и общее количество просроченных платежей
debtors_total_income = data.groupby('total_income_category')['debt'].agg(['count', 'mean', 'sum'])

#преобразовываем среднее значение в проценты и округляем до двух знаков после точки
debtors_total_income['mean'] = (debtors_total_income['mean'] * 100).round(2)

#сортируем по общему количествту должников по убыванию
debtors_total_income = debtors_total_income.sort_values(['count'], ascending=False)

debtors_total_income



count  mean  sum
total_income_category                  
очень высокий           9777  7.76  759
высокий                 9289  8.60  799
средний                 2265  7.68  174

<div style='border:solid green 2px; pdding:20px'>
    
   **Вывод:** 
- Большинство должников (около 90%) имеют средний или выше среднего уровень дохода.
- Средняя сумма задолженности заемщиков с низким уровнем дохода (9.09) выше, чем у заемщиков со средним уровнем дохода (7.67). Это может быть связано с низкой кредитоспособностью. 
- Заемщики со средним и с очень высоким уровнем дохода оплачивают кредиты в срок, чем те, у кого высокий и низкий уровень дохода.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

<div class='alert alert-info'>Для выявления влияния разных кредитов на его возврат в срок группируем данные по целям кредита ('purpose_category), затем применяем агрегирущие функции к столбцу 'debt': считаем количество должников в каждой группе по целям кредита, затем средние значения задолженности по целям кредита и общую сумма долга в каждой группе по цели кредита.</div>

In [46]:
#группируем данные по целям кредита ('purpose_category), затем применяем агрегирущие функции к столбцу 'debt': 
#считаем количество должников в каждой группе по целям кредита, 
#затем средние значения задолженности по целям кредита,
#и на общую сумма долга в каждой группе по цели кредита
debt_purpose = data.groupby('purpose_category')['debt'].agg(['count', 'mean', 'sum'])

#преобразовываем среднее значение в проценты и округляем до двух знаков после точки
debt_purpose['mean'] = (debt_purpose['mean'] * 100).round(2)

#сортируем по общему количествту должников по убыванию
debt_purpose = debt_purpose.sort_values(['count'], ascending=False)

debt_purpose

count  mean  sum
purpose_category                          
операции с недвижимостью  10751  7.26  780
операции с автомобилем     4279  9.35  400
получение образования      3988  9.25  369
проведение свадьбы         2313  7.91  183

<div style='border:solid green 2px; pdding:20px'>
 
**Вывод**: 
- Наибольшее количество кредитов приходится на операции с недвижимостью (10751 кредит).
- Наименьшей популярностью пользуются кредиты на проведение свадеб (2313 кредит). 
- Самый высокий процент просроченных кредитов - операции с автомобилем (9.35%).
- Категория 'получение образования' также является одним из просрачиваемых кредитов (9.25%)
- Самый низкий процент просроченных кредитов приходится на операции с недвижимостью (7.26%). 
- На операции с недвижимостью больше всего просроченных платежей (780).


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

<div class='alert alert-info'>
    
**Ответ:** 
Конкретно в этих данных пропущенные значения у пенсионера в столбце 'total_income'. У них нет зарплаты, соответственно
тут респондент специально не заполнил данные. Также в этой же строке в столбце 'days_employed' пропущенное 
значение тоже неслучайно, ведь у пенсионера нет отработанных дней. 
Пропуск в 'total_income' и 'days_employed' заполнен медианным значением столбца. 

    
    
**Возможные причины появления пропусков в данных**:
- ошибки в сборе данных
- технические проблемы: сбой системы и прочее
- потеря данных
- нежелание респондентов заполнять данные
- недостаточный объем выборки </div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

<div class='alert alert-info'>
    
**Ответ:** Медиана, в отличие от средней арифметической, не искажает факты. Медиана является более устойчивой мерой центральной
тенденции. Например, среднюю зарплату сотрудников лучше всего считать медианой, так как у всех она может быть разная. 
Также среднюю зарплату в стране истинно покажет медиана, так как разброс в доходах у населения сильно отличается. 

### Шаг 4: общий вывод.

<div class='alert alert-info'>
    
**В данных были обнаружены следующие ошибки**:

- Пропущены значения в двух столбцах: 'days_employed' (2174) и 'total_income' (2174)
- Пропуски заменены на медианное значение среди всех записей с тем же типом.
- В столбце 'days_employed' были записи в отрицательном значении: заменены методом abs()
- Столбец 'total_income' из строкового типа заменен на целочисленный методом astype(). Используем этот метод, чтобы получить целое число, а методом numeric() получим вещественное.
- Всего найдено 71 дубликат (метод duplicated()). Удалены методом drop_duplicates()


<div style='border:solid green 2px; pdding:20px'>
    
**На основании проведенного анализа можно сделать такие выводы**:
- Средняя сумма задолженности заемщиков без детей (7.54) ниже, чем у заемщиков с детьми (9.30 для малодетных и 8.16 для многодетных). Это может быть связано с тем, что заемщики без детей могут иметь более высокий уровень доходов и, следовательно, могут брать кредиты на более крупные суммы, чем заемщики с детьми.
- Семейное положение может влиять на оплату кредита в срок. Количество женатых должников значительно больше, чем других категорий (12261), гражданский брак	(4134), неженатые (2796), в разводе (1189). Процент просрочки кредита больше у неженатых и незамужних должников (9.76%). Может быть связано с тем, что люди в браке чаще имеют больше обязательств, берут ответсвтвенность за всех членов семьи, лучше кредитная история и финансовая грамотность.
- Общая сумма задолженности заемщиков без детей (1063) выше, чем у заемщиков с детьми (638 для малодетных и 31 для многодетных). Это может быть связано с тем, что заемщики без детей могут брать кредиты на более крупные суммы, чем заемщики с детьми.
- Среди заемщиков с детьми, малодетные семьи имеют более высокий уровень задолженности, чем многодетные семьи. Тут может играть роль то, что малодетные семьи могут иметь более высокий доход и, следовательно, брать кредиты на крупные суммы, чем многодетные. 
- Самая популярная цель кредитов связана с операциями с недвижимостью (10751). Ипотеки и рассрочки могут иметь низкий процент и ежегодное повышение роста цен на недвижимость заставляет людей более охотно брать деньги в долг на эту категорию.
- Цель кредита также имеет влияние на оплату кредита в срок. Должники охотнее оплачивают срок операции, связанные с недвижимостью (7.26%), чем с покупкой автомобиля (9.35%). Это может быть связано с тем, что недвижимость является более ценным активом, чем покупка автомобиля. Заемщики больше мотивированы платить за дом/квартиру вовремя из-за боязни потерять имущество и повышения ставки. 

Эти выводы могут быть полезны для банков и кредитных организаций. Но эти данные не могут быть абсолютно точными, так как для более точного анализа потребуются дополнительные данные для оценки кредитоспособности заемщиков. На выводы могут повлиять кредитная история заемщика, дополнительные источники дохода, уровень обязательность заемщиков и прочие детали. 